### ISSUE TITLE
Filter for product variations list

### ISSUE THREAD

**Issue by joshuatf** [woothemes/woocommerce#35792]
Users can filter the list by one or multiple variations and several options.

**Design**

BaoQ3PGPnhAiJCsJtxgej3-fi-106%3A3188&t=Jsh233onKmJpJYtN-0

![Image](https://user-images.githubusercontent.com/10561050/204899288-e317c81f-b280-4de4-8777-5b88778847dd.png)

**Acceptance criteria**

 * [ ] When an option is added, a filter should be added to the top of the variations list with possible values
 * [ ] If many filters exist, they're displayed as rows (other elements in the table header are then aligned top). `BaoQ3PGPnhAiJCsJtxgej3-fi-307%3A155754&t=ldR14Ed8fjD5K2wD-4`
 * [ ] Selecting one or more filters should filter the list should filter the variations by the selected filters
 * [ ] Pagination should update to reflect filtered items and not the total item count
 * [ ] Variation data store should allow filter by the `local_attributes` property

**Comment by joshuatf** [comment]
@woocommerce/mothra before working on this issue, I think we should verify if there is a way to retrieve filtered variations from the REST endpoints.  While we can filter client-side, pagination will not work as expected if we do this.

I see a `search` option that may or may not work for this case in the [variations endpoint](https://woocommerce.github.io/woocommerce-rest-api-docs/#product-variations).  We should check to see if that will work and if not, create a separate issue and estimate to determine how difficult it would be to add in filtering to those endpoints.

@jarekmorawski I'm also curious what would happen if many filters exist.  Would the dropdown filters collapse down or get truncated in some way?

**Comment by jarekmorawski** [comment]
> I'm also curious what would happen if many filters exist. Would the dropdown filters collapse down or get truncated in some way?

That was my initial approach, but it'd require us to build a new component for handling nested dropdowns, wouldn't it? To avoid that, we could show the filters in multiple rows. I've updated the ACs to reflect that.

<img width="565" alt="image" src="https://user-images.githubusercontent.com/79307566/204986430-e6a82508-743a-43b1-b66a-475fa7a12e02.png">

**Comment by joshuatf** [comment]
On further investigation, it looks like the product variations controller uses the products controller underneath.  This contains an optional parameter to filter by an attribute.

https://github.com/woocommerce/woocommerce/blob/884d3f4237fa03f5c435d02edc3a9567d06a9c2f/plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-products-v2-controller.php#L2341-L2352

However, this will only allow us to filter by a single option (attribute).  I see 3 options going forward:

1. Use a single attribute filter only.  If one attribute has already been selected, we could hide or disable the others so that multiple filters can't be selected.
2. Punt this issue to the next cycle and look at extending the product controller to allow for filtering by multiple attributes.
3. Fetch all variations and handle filtering client-side.  This means that we have to handle pagination client-side and sites with a lot of variations will not be as performant.

@jarekmorawski do you have a preferred path forward from above? cc @octaedro @mdperez86 @mattsherman

**Comment by mattsherman** [comment]
> Punt this issue to the next cycle and look at extending the product controller to allow for filtering by multiple attributes.

I'd favor this option. In general, when we can move functionality server-side, let's do it. It allows extensions to make use of this filtering as well, in addition to the mobile apps.

> Use a single attribute filter only.

This seems too limiting. I can see merchants wanting to filter by multiple attributes.

**Comment by jarekmorawski** [comment]
> Punt this issue to the next cycle and look at extending the product controller to allow for filtering by multiple attributes.

I agree. We should support filtering by multiple attributes out of the box, and it's better not to have any than something that won't be useful to the users. In that case, we shouldn't show the filters at all and leave that space empty (we'd still show there the list control buttons when at least one item is selected).

**Comment by mdperez86** [comment]
I always like to avoid client side pagination and filtering. As you said @joshuatf this could introduce performance issues and also does not scale in time. So I prefere server side pagination + filtering + ordering and heavy operations.

**Comment by octaedro** [comment]
Great analysis @joshuatf,

I'm in favor of option 2 -> `Punt this issue to the next cycle and look at extending the product controller to allow for filtering by multiple attributes.`

**Comment by joshuatf** [comment]
Thanks for the feedback everyone!  I've moved this back to prioritization for now.

Before beginning work on this issue, we should create another issue to tackle adding multiple attribute filtering to the REST API.

**Comment by pmcpinto** [comment]
> That was my initial approach, but it'd require us to build a new component for handling nested dropdowns, wouldn't it? To avoid that, we could show the filters in multiple rows. I've updated the ACs to reflect that.

@jarekmorawski do you think this approach can scale if a store uses +10 or 15 attributes in a product? This is more like an edge case, so it doesn't need to be solved right now, but we should keep it in mind.

**Comment by mattsherman** [comment]
> On further investigation, it looks like the product variations controller uses the products controller underneath. This contains an optional parameter to filter by an attribute.

> Before beginning work on this issue, we should create another issue to tackle adding multiple attribute filtering to the REST API.

A few technical details regarding the current query support for production variations by attribute. I learned all of these when I was playing around with the API in Postman, trying to understand it better. Hopefully this will saves others a little time when digging into things, as some of these details definitely tripped me up.

- `/wp-json/wc/v3/products/<product_id>/variations`
    - does not currently support the `attribute` or `attribute_term` query params... they are ignored

- `/wp-json/wc/v3/products`
    - does support the `attribute` and `attribute_term` query params (as Josh mentioned above)
    - both query params need to be provided, otherwise they are ignored
    - the `attribute` param needs to match an existing produce attribute slug, otherwise the param is ignored (**if you pass in a non-existent slug, it returns all results!**)
        - the slug is *not* the slug you see in the UI... it is the slug prefixed by `pa_`; so, to search for an attribute with a slug of `global-attribute-1`, you need to provide `pa_global-attribute-1`
    - the `attribute_term` param is the `id` for the term, which you can find via the `/wp-json/wc/v3/products/attributes/<attribute_id>/terms` endpoint
    - I am pretty sure these params only support global attributes; **I do not think there is a way to query by local attributes**

**Comment by joshuatf** [comment]
Once https://github.com/woocommerce/woocommerce/pull/36201 is merged, this will be unblocked and ready to work on.

Estimate: 3

**Comment by joshuatf** [comment]
I'm going to unassign myself this for the time being so that we can focus on bugs in the newly added feature set before adding new features.

**Comment by jarekmorawski** [comment]
Tackled differently in the new block-based design.
